src/parser.js

```diff
@@ -1561,24 +1561,18 @@ export class GenericParser extends Tokenizer {
           // the maximally obnoxious case: `async (`
           let afterAsyncStartState = this.startNode();
           let previousAwaitLocation = this.firstAwaitLocation;
+          let preParseParamState = this.saveLexerState();
           this.firstAwaitLocation = null;
-          let { args, locationFollowingFirstSpread } = this.parseArgumentList();
+          let { args } = this.parseArgumentList();
           if (this.isBindingElement && !this.hasLineTerminatorBeforeNext && this.match(TokenType.ARROW)) {
-            if (locationFollowingFirstSpread !== null) {
-              throw this.createErrorWithLocation(locationFollowingFirstSpread, ErrorMessages.UNEXPECTED_TOKEN(','));
-            }
+            this.restoreLexerState(preParseParamState);
+            let { items, rest } = this.parseParams();
```

Comment:
We have generally designed the parser to avoid backtracking - i.e., we never parse the same arbitrarily-long piece of input twice. With this strategy, we end up parsing the part between the parentheses in `async (x = function(){abitraryilyComplexCode()}) => {}` twice. That's bad; I'd like to find a different strategy to solve this.

Reply:
Fixed